Added support for using PAM with basic-auth.

arimed opened the pull request:
@erictune If basic-auth is used for authentication when a client communicates with the api server
and the BASIC_AUTH_PAM env var is on, then the configured PAM module is used to validate the
user name/password pair.  For example the local PAM module may be configured to lookup the user name/password in Active Directory.

Commit:
Added support for using PAM with basic-auth.  If basic-auth is used
for authentication when a client communicates with the api server and
the BASIC_AUTH_PAM env var is on, then the configured pam module is
used to validate the user name/password pair.  For example the local
PAM module may be configured to lookup the user name/password in
Active Directory.

liggitt commented on plugin/pkg/auth/authenticator/request/basicauth/basicauth.go:
```diff
@@ -33,11 +36,52 @@ func New(auth authenticator.Password) *Authenticator {
 	return &Authenticator{auth}
 }
 
+type Credentials struct {
```
these belong in a separate file, implemented as a authenticator.Password. See packages under https://github.com/kubernetes/kubernetes/tree/master/plugin/pkg/auth/authenticator/password

liggitt commented on hack/lib/golang.sh:
```diff
@@ -397,7 +397,7 @@ kube::golang::build_binaries_for_platform() {
     fi
     if [[ "${#statics[@]}" != 0 ]]; then
-      CGO_ENABLED=0 go install -installsuffix cgo "${goflags[@]:+${goflags[@]}}" \
+      CGO_ENABLED=1 go install -installsuffix cgo "${goflags[@]:+${goflags[@]}}" \
```
not sure what the implications of this change are, but they might be big

liggitt replied:
I wouldn't expect to need this change. Assuming there are platform dependencies, add build tags to the PAM password impl to limit dependencies to the PAM lib to linux (or whatever other platforms support it) and stub the impl for other platforms

arimed replied:
Pam uses a C lib, so CGO must be enabled.

liggitt replied:
Doesn't building with cgo make it link to libraries dynamically, not statically? If so, that is reversing the intent of this part of the build script

liggitt replied:
https://github.com/kubernetes/kubernetes/commit/9a901ed46db225f588cc691feeb47e7ace92bbc8 explicitly made kube-apiserver a static binary. If we want to reverse that to add a cgo dependency, we'll need broader agreement

arimed replied:
without CGO_ENABLED=1, I get the following build error:
icauth/basicauth.go:22:2: C source files not allowed when not using cgo or SWIG: transaction.c

liggitt replied:
Is the lib cross platform? If not, you'll need to separate out the bits that are platform dependent (or call a platform dependent package) into a separate file, and use build tags to include them selectively, then create a separate file the implements the same functions and use build tags to select that file for all the unsupported platforms.

liggitt commented:
rather than triggering off envvars, the PAM impl should be a particular password authenticator that could be plugged into the basic auth request authenticator

liggitt commented on plugin/pkg/auth/authenticator/request/basicauth/basicauth.go:
```diff
@@ -33,11 +36,52 @@ func New(auth authenticator.Password) *Authenticator {
 		return nil, false, nil
 	}
+
+	if os.Getenv("BASIC_AUTH_PAM") == "1" {
```
All of these changes belong in a PAMPassword impl and package

liggitt commented on plugin/pkg/auth/authenticator/request/basicauth/basicauth.go:
```diff
@@ -17,10 +17,13 @@ limitations under the License.
-
+	"errors"
+	"github.com/golang/glog"
+	"github.com/msteinert/pam"
```
All packages have to be vendored in godeps. Add this package in a separate commit

arimed commented:
I just submitted a CLA for Centrify corp

erictune commented:
@lavalamp or @thockin to comment on the proposal here to enable CGO.

liggitt commented on plugin/pkg/auth/authenticator/request/basicauth/basicauth.go:
```diff
@@ -33,11 +36,52 @@ func New(auth authenticator.Password) *Authenticator {
+			Password: password,
+		}
+
+		module := os.Getenv("KUBE_PAM_MODULE_NAME")
```
enabling/specifying the PAM module would be an API server flag, the presence of which would add a PAM password authenticator to the auth chain

liggitt replied:
See
cmd/kube-apiserver/app/options/options.go
cmd/kube-apiserver/app/server.go
pkg/apiserver/authenticator/authn.go

arimed replied:
ok, but to collect the user/password via the browser, you would still need basic auth turned on right?

liggitt replied:
In general, basic auth from browsers against APIs is a really bad idea. it opens users up to CSRF, which means you need to add CSRF protection in front of the API, which means non-browser clients have to deal with CSRF protection. This plumbing will recognize credentials sent using basic auth, but it won't send basic auth challenges, so a browser wouldn't be prompted to enter a username/password.

arimed replied:
agreed.  Note that in the current implementation (without my changes), if basic auth option is turned on for api server, then the browser does prompt the user for username & password.

liggitt replied:
That is currently only for the toy basic auth file implementation. We should not extend that behavior to real password impls like this.

liggitt replied:
See threads in https://github.com/kubernetes/kubernetes/pull/6508#issuecomment-90622897

Commit:
Incorporated feedback from the initial review.  Made PAM auth a seperate
module such that it can be configured via an apiserver option to be used
in conjunction with basic auth.  I did not change the CGO flag, as Eric
solicited input on this topic.  Also, the CLA should be in completed state
now.
@erictune

Commit:
forgot to add the pampassword.go (that implements the new pam module)
as part of my last checkin.

liggitt commented on plugin/pkg/auth/authenticator/request/basicauth/basicauth.go:
```diff
@@ -17,10 +17,9 @@ limitations under the License.
 package basicauth
 
 import (
-	"net/http"
```
Undo this change

arimed replied:
will do, one sec

liggitt commented on pkg/apiserver/authenticator/authn.go:
```diff
@@ -59,6 +61,14 @@ func New(config AuthenticatorConfig) (authenticator.Request, error) {
 	}
 
+	if config.PamPassword {
+		basicAuth, err := newAuthenticatorFromPamPassword(config.BasicAuthFile)
```
Should be PamPassword, right?

liggitt commented on pkg/apiserver/authenticator/authn.go:
```diff
@@ -44,6 +45,7 @@ type AuthenticatorConfig struct {
 	ServiceAccountLookup      bool
 	ServiceAccountTokenGetter serviceaccount.ServiceAccountTokenGetter
 	KeystoneURL               string
+	PamPassword               bool
```
Make this a string that holds the name of the service, not a bool

liggitt commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -374,7 +375,7 @@ func Run(s *options.APIServer) error {
 			PublicAddress:             s.AdvertiseAddress,
 			Authenticator:             authenticator,
-			SupportsBasicAuth:         len(s.BasicAuthFile) > 0,
+			SupportsBasicAuth:         len(s.BasicAuthFile) > 0 || s.PamPassword,
```
This flag controls sending basic auth challenges, which will prompt browsers for  passwords, without any way to log out, and without any cross site request forgery protection. We should not enable this flag for more password authentication methods without addressing those issues

liggitt commented on cmd/kube-apiserver/app/options/options.go:
```diff
@@ -96,6 +96,7 @@ type APIServer struct {
 	SSHKeyfile                 string
 	MaxConnectionBytesPerSec   int64
 	KubernetesServiceNodePort  int
+	PamPassword                bool
```
Also a string that holds the PAM service name

liggitt commented on plugin/pkg/auth/authenticator/password/pampassword/pampassword.go:
```diff
@@ -0,0 +1,76 @@
+	return "", errors.New("unexpected")
+}
+
+// NewAllow returns a password authenticator that allows any non-empty username
```
Fix godoc

liggitt replied:
Take the PAM module name here and persist it in the struct, and use it instead of reading an envvar in AuthenticatePassword

liggitt commented on plugin/pkg/auth/authenticator/password/pampassword/pampassword.go:
```diff
@@ -0,0 +1,76 @@
+		Password: password,
+	}
+
+	module := os.Getenv("KUBE_PAM_MODULE_NAME")
```
Read the module from the struct. I don't know enough about PAM to know if a default like this is expected or not, but I'd rather see it be explicitly specified

liggitt commented on plugin/pkg/auth/authenticator/password/pampassword/pampassword.go:
```diff
@@ -0,0 +1,76 @@
+
+// NewAllow returns a password authenticator that allows any non-empty username
+func NewPamPassword() authenticator.Password {
+	return pamPaswdAuthenticator{}
```
This is the prime file for splitting by platform. Have a pampassword_supported.go that contains this impl, and a pampassword_unsupported.go that contains a NewPAMPassword() func that returns `nil, errors.New("PAM password authentication is not supported")`
